src/coreclr/jit/gentree.cpp

```diff
@@ -14411,44 +14412,46 @@ GenTree* Compiler::gtFoldExprSpecial(GenTree* tree)
+                    if (op->IsBoxedValue())
+                    {
+                        // The tree under the box must be side effect free
+                        // since we will drop it if we optimize.
+                        assert(!gtTreeHasSideEffects(op->AsBox()->BoxOp(), GTF_SIDE_EFFECT));
 
-                    // See if we can optimize away the box and related statements.
-                    GenTree* boxSourceTree = gtTryRemoveBoxUpstreamEffects(op);
-                    bool     didOptimize   = (boxSourceTree != nullptr);
+                        // See if we can optimize away the box and related statements.
+                        wrapEffects = (gtTryRemoveBoxUpstreamEffects(op) == nullptr);
+                    }
 
-                    // If optimization succeeded, remove the box.
-                    if (didOptimize)
+                    // Set up the result of the compare.
+                    int compareResult;
+                    if (oper == GT_GT)
                     {
-                        // Set up the result of the compare.
-                        int compareResult = 0;
-                        if (oper == GT_GT)
-                        {
-                            // GT_GT(null, box) == false
-                            // GT_GT(box, null) == true
-                            compareResult = (op1 == op);
-                        }
-                        else if (oper == GT_EQ)
-                        {
-                            // GT_EQ(box, null) == false
-                            // GT_EQ(null, box) == false
-                            compareResult = 0;
-                        }
-                        else
-                        {
-                            assert(oper == GT_NE);
-                            // GT_NE(box, null) == true
-                            // GT_NE(null, box) == true
-                            compareResult = 1;
-                        }
-
-                        JITDUMP("\nSuccess: replacing BOX(valueType) %s null with %d\n", GenTree::OpName(oper),
-                                compareResult);
+                        // GT_GT(null, op) == false
+                        // GT_GT(op, null) == true
+                        compareResult = (op1 == op);
+                    }
+                    else if (oper == GT_EQ)
+                    {
+                        // GT_EQ(op, null) == false
+                        // GT_EQ(null, op) == false
+                        compareResult = 0;
+                    }
+                    else
+                    {
+                        assert(oper == GT_NE);
+                        // GT_NE(op, null) == true
+                        // GT_NE(null, op) == true
+                        compareResult = 1;
+                    }
 
-                        return NewMorphedIntConNode(compareResult);
+                    GenTree* newTree = gtNewIconNode(compareResult);
+                    if (wrapEffects)
```

Comment:
just wrap unconditionally, `gtWrapWithSideEffects` won't create a COMMA if there are no side-effects

Reply:
AFAIR we can't use the `op` if `gtTryRemoveBoxUpstreamEffects` removes the box, that's why I did the check.

Reply:
not sure I understand, if it removes - does it leave a tree with a side-effect?

Reply:
> not sure I understand, if it removes - does it leave a tree with a side-effect?

As far as @SingleAccretion explained it to me on Discord, using the tree when `gtTryRemoveBoxUpstreamEffects` succeedes is nonsensical since the original tree is invalid due to the earlier box not existing anymore.